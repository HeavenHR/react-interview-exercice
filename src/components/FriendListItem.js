import React, { Component } from "react";
import PropTypes from 'prop-types';
import classnames from "classnames";
import styles from "./FriendListItem.css";
import * as genders from "../constants/GenderTypes";

class FriendListItem extends Component {
  render() {
    return (
      <li className={styles.friendListItem} name='friend-item'>
        <div className={styles.friendInfos}>
          <div>
            <span>{this.props.name} </span>
          </div>
          <div>
            <small>
              {this.printGender(this.props.name, this.props.gender)}{" "}
            </small>
          </div>
          <div>
            <small>xx friends in common</small>
          </div>
        </div>
        <div className={styles.friendActions}>
          <button
            className={`btn btn-default ${styles.btnAction}`}
            onClick={() => this.props.starFriend(this.props.id)}
          >
            <i
              className={classnames("fa", {
                "fa-star": this.props.starred,
                "fa-star-o": !this.props.starred
              })}
            />
          </button>
          <button
            className={`btn btn-default ${styles.btnAction}`}
            onClick={() => this.props.deleteFriend(this.props.id)}
          >
            <i className="fa fa-trash" />
          </button>
        </div>
      </li>
    );
  }

  printGender(name, gender) {
    if (gender === genders.GENDER_TYPE.MALE) {
      return <span>[MALE]</span>;
    } else if (gender === genders.GENDER_TYPE.FEMALE) {
      return <span>[FEMALE]</span>;
    } else {
      return <span>GENDER WAS NOT SELECTED</span>;
    }
  }
}

FriendListItem.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  gender: PropTypes.string.isRequired,
  starred: PropTypes.bool,
  starFriend: PropTypes.func.isRequired
};

export default FriendListItem;
