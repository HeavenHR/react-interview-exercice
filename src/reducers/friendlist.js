import {random} from 'lodash'
import * as types from '../constants/ActionTypes';

const initialState = {
  friendsById: [
    {
      id:random(10000,1000000,false),
      name: 'Theodore Roosevelt',
      gender: 'male',
      starred: true
    },
    {
      id:random(10000,1000000,false),
      name: 'Abraham Lincoln',
      gender: 'male',
      starred: false
    },
    {
      id:random(10000,1000000,false),
      name: 'George Washington',
      gender: 'male',
      starred: false
    }
  ]
};

export default function friends(state = initialState, action) {
  switch (action.type) {
    case types.ADD_FRIEND:
      return {
        ...state,
        friendsById: [
          ...state.friendsById,
          {
            name: action.friend.name,
            gender: action.friend.gender,
            id:random(10000,1000000,false),
          }
        ],
      };
    case types.DELETE_FRIEND:
      return {
        ...state,
        friendsById: state.friendsById.filter((item) => item.id !== action.id)
      };
    case types.STAR_FRIEND:
      let friends = [...state.friendsById];
      let friend = friends.find((item) => item.id === action.id);
      friend.starred = !friend.starred;
      return {
        ...state,
        friendsById: friends
      };

    default:
      return state;
  }
}
