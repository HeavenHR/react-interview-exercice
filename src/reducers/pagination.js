import * as types from '../constants/ActionTypes';

const initialState = {
  currentPageNo: 0,
  pageSize: 2
};

export default function pagination(state = initialState, action) {
  switch (action.type) {

    case types.UPDATE_AFTER_CHANGE:
      const {currentPageNo} = action;
      return {
        ...state,
        currentPageNo
      };

    case types.SET_PAGE:
      return {
        ...state,
        currentPageNo: action.pageNo
      };

    default:
      return state;
  }
}
