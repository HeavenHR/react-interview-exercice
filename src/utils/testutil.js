import { mount, shallow } from 'enzyme';
import { prettyPrint } from 'html';
import * as types from '../constants/ActionTypes';

export const mockFunction = jest.fn(value => value);

export const sampleFriend = {
  name: 'Sample friend',
  gender: 'male'
};

export const sampleId = 1;

export const createMountedComponent = ({ children, isShallow }) => {
  const mountedComponent = isShallow ? shallow(children) : mount(children);
  mountedComponent.htmlRender = () => prettyPrint(mountedComponent.html(), { indent_size: 2 });

  return mountedComponent;
};

export const inputProps = {
  value: 'sample name',
  onSubmit: mockFunction,
  onChange: mockFunction,
  isInvalid: false,
  errorMessage: 'Please fill friend name'
};

export const buttonGroupProps = {
  selected: '',
  options: [
    { value: 'male', showIcon: true },
    { value: 'female', showIcon: true }
  ],
  onClick: mockFunction,
  isInvalid: false,
  errorMessage: 'Please select gender'
};

export const changeEventObject = { target: { value: 'sample text' } };

export const samplePages = [1, 2, 3, 4, 5, 6, 7];

export const totalPages = samplePages.length;

export const addFriendAction = {
  type: types.ADD_FRIEND,
  friend: sampleFriend
};

export const expectedAddFriendState = {
  friendsById: [
    {
      name: 'Sample friend',
      gender: 'male'
    },
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const deleteFriendAction = {
  type: types.DELETE_FRIEND,
  id: 0
};

export const expectedDeleteFriendState = {
  friendsById: [
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const starFriendAction = {
  type: types.STAR_FRIEND,
  id: 1
};

export const expectedStarFriendState = {
  friendsById: [
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: true,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const updatePaginationAction = {
  type: types.UPDATE_PAGINATION,
  currentPage: 2
};

export const expectedUpdatedPaginationState = {
  pageSize: 2,
  currentPage: 2,
  pagingRange: 3
};
