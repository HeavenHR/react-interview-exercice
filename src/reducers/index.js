import { combineReducers } from 'redux';
import { default as friendlist } from './friendlist';
import { pagination } from './pagination';

export default combineReducers({ friendlist, pagination });
