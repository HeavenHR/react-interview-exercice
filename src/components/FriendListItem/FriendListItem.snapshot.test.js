import React from 'react';

import FriendListItem from './FriendListItem';
import { createMountedComponent } from '../../utils/testutil';
import { initialState } from '../../reducers/friendlist';

describe('<FriendListItem /> should', () => {
  const mountedFriendListItem = createMountedComponent({
    children: (
      <FriendListItem
        id={0}
        {...initialState.friendsById[0]}
        {...localStorage.actions}
      />)
  });

  const buttons = mountedFriendListItem.find('button').map(item => item);

  it('render correctly according to props', () => {
    expect(mountedFriendListItem.htmlRender()).toMatchSnapshot();
  });

  it('simulate starFriend action when click star button', () => {
    buttons[0].simulate('click');

    expect(mountedFriendListItem.props().starFriend).toBeCalledWith(0);
  });

  it('simulate delete action when click delete button', () => {
    buttons[1].simulate('click');

    expect(mountedFriendListItem.props().deleteFriend).toBeCalledWith(0);
  });
});
