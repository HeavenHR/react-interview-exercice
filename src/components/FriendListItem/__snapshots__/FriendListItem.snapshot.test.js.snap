exports[`<FriendListItem /> should render correctly according to props 1`] = `
"<li class=\"list-group-item undefined\">
  <div>
    <div><span>Theodore Roosevelt</span>
    </div>
    <div><i class=\"fa fa-male\"></i><small>male</small>
    </div>
    <div><small>xx friends in common</small>
    </div>
  </div>
  <div>
    <button class=\"btn btn-default undefined\"><i class=\"fa fa-star\"></i>
    </button>
    <button class=\"btn btn-default undefined\"><i class=\"fa fa-trash\"></i>
    </button>
  </div>
</li>"
`;
